system-contracts/test/MsgValueSimulator.spec.ts

```diff
@@ -0,0 +1,194 @@
+import type { L2EthToken, MockContract, MsgValueSimulator } from "../typechain";
+import { L2EthTokenFactory, MsgValueSimulatorFactory } from "../typechain";
+import { deployContract, deployContractOnAddress, getWallets, loadZasmBytecode, setCode } from "./shared/utils";
+import {
+  TEST_BOOTLOADER_FORMAL_ADDRESS,
+  TEST_ETH_TOKEN_SYSTEM_CONTRACT_ADDRESS,
+  TEST_MSG_VALUE_SYSTEM_CONTRACT_ADDRESS,
+} from "./shared/constants";
+import { prepareEnvironment } from "./shared/mocks";
+import { expect } from "chai";
+import { EXTRA_ABI_CALLER_ADDRESS, encodeExtraAbiCallerCalldata } from "./shared/extraAbiCaller";
+import { BigNumber } from "ethers";
+import { Contract } from "zksync-web3";
+import * as hardhat from "hardhat";
+
+describe("MsgValueSimulator tests", () => {
+  let messageValueSimulator: MsgValueSimulator;
+  let extraAbiCaller: Contract;
+  let mockContract: MockContract;
+  let l2EthToken: L2EthToken;
+  const wallet = getWallets()[0];
+  const msgSender = "0x000000000000000000000000000000000000beef";
+
+  before(async () => {
+    // Prepare environment
+    await prepareEnvironment();
+    await deployContractOnAddress(TEST_MSG_VALUE_SYSTEM_CONTRACT_ADDRESS, "MsgValueSimulator");
+    await deployContractOnAddress(TEST_ETH_TOKEN_SYSTEM_CONTRACT_ADDRESS, "L2EthToken");
+    mockContract = (await deployContract("MockContract")) as MockContract;
+
+    // Prepare MsgValueSimulator
+    messageValueSimulator = MsgValueSimulatorFactory.connect(TEST_MSG_VALUE_SYSTEM_CONTRACT_ADDRESS, wallet);
+    const extraAbiCallerBytecode = await loadZasmBytecode("ExtraAbiCaller", "test-contracts");
+    await setCode(EXTRA_ABI_CALLER_ADDRESS, extraAbiCallerBytecode);
+    extraAbiCaller = new Contract(EXTRA_ABI_CALLER_ADDRESS, [], wallet);
+
+    // Prepare ETH token
+    l2EthToken = L2EthTokenFactory.connect(TEST_ETH_TOKEN_SYSTEM_CONTRACT_ADDRESS, wallet);
```

Comment:
I think it may be easier to mock the token contract, usually we are mocking other contracts. Don't you think so?

Reply:
I wanted to avoid compile time bloat.
Now that you suggested it, I tried to do it but I believe [this line](https://github.com/neotheprogramist/era-contracts/blob/df268ac8d96b647b400406e65a834a8a6520716b/system-contracts/contracts/MsgValueSimulator.sol#L46-L47) makes it impossible.

Reply:
Well, not if I deployed on the same address but this could be a problem for the other tests.  Also I don't think it would be much easier.

Let me know if I should do it anyway

Reply:
I would do it, at least it will be consistent with other tests, I think the lines you mentioned above is not a problem, you just need properly mock the contract. But why do you think it may be problem for the other tests?

Reply:
Only because it would be at the same address as the real `TEST_ETH_TOKEN_SYSTEM_CONTRACT_ADDRESS` in other tests.